subsys/net/ip/net_pkt.c

```diff
@@ -2176,6 +2188,472 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+#else /* !CONFIG_NET_BUF_FIXED_DATA_SIZE */
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout,
+					const char *caller, int line)
+#else
+static struct net_buf *pkt_alloc_buffer(struct net_buf_pool *pool,
+					size_t size, s32_t timeout)
+#endif
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc_len(pool, size, timeout);
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+	NET_FRAG_CHECK_IF_NOT_IN_USE(buf, buf->ref + 1);
+
+	net_pkt_alloc_add(buf, false, caller, line);
+
+	NET_DBG("%s (%s) [%d] frag %p ref %d (%s():%d)",
+		pool2str(pool), get_name(pool), get_frees(pool),
+		buf, buf->ref, caller, line);
+#endif
+
+	return buf;
+}
+
+#endif /* CONFIG_NET_BUF_FIXED_DATA_SIZE */
+
+static size_t pkt_buffer_length(struct net_pkt *pkt,
+				size_t size,
+				enum net_ip_protocol proto,
+				size_t existing)
+{
+	size_t max_len = net_if_get_mtu(net_pkt_iface(pkt));
+	sa_family_t family = net_pkt_family(pkt);
+
+	/* Family vs iface MTU */
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		max_len = max(max_len, NET_IPV6_MTU);
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		max_len = max(max_len, NET_IPV4_MTU);
+	}
+
+	max_len -= existing;
+	max_len = min(size, max_len);
+
+	return min(size, max_len);
+}
+
+static size_t pkt_headers_buffer_length(struct net_pkt *pkt,
+					sa_family_t family,
+					enum net_ip_protocol proto)
+{
+	size_t hdr_len = 0;
+
+	if (family == AF_UNSPEC) {
+		return  0;
+	}
+
+	/* Family header */
+	if (IS_ENABLED(CONFIG_NET_IPV6) && family == AF_INET6) {
+		hdr_len += NET_IPV6H_LEN;
+	} else if (IS_ENABLED(CONFIG_NET_IPV4) && family == AF_INET) {
+		hdr_len += NET_IPV4H_LEN;
+	}
+
+	/* + protocol header */
+	if (IS_ENABLED(CONFIG_NET_TCP) && proto == IPPROTO_TCP) {
+		hdr_len += NET_TCPH_LEN + NET_TCP_MAX_OPT_SIZE;
+	} else if (IS_ENABLED(CONFIG_NET_UDP) && proto == IPPROTO_UDP) {
+		hdr_len += NET_UDPH_LEN;
+
+		if (IS_ENABLED(CONFIG_NET_RPL_INSERT_HBH_OPTION)) {
+			hdr_len += NET_RPL_HOP_BY_HOP_LEN;
+		}
+	} else if (proto == IPPROTO_ICMP || proto == IPPROTO_ICMPV6) {
+		hdr_len += NET_ICMPH_LEN;
+	}
+
+	NET_DBG("HDRs length estimation %zu", hdr_len);
+
+	return hdr_len;
+}
+
+static size_t pkt_get_size(struct net_pkt *pkt)
+{
+	struct net_buf *buf = pkt->buffer;
+	size_t size = 0;
+
+	while (buf) {
+		size += buf->size;
+		buf = buf->frags;
+	}
+
+	return size;
+}
+
+
+#if NET_LOG_LEVEL >= LOG_LEVEL_DBG
+int net_pkt_alloc_buffer_debug(struct net_pkt *pkt,
+			       size_t size,
+			       enum net_ip_protocol proto,
+			       s32_t timeout,
+			       const char *caller,
+			       int line)
+#else
+int net_pkt_alloc_buffer(struct net_pkt *pkt,
+			 size_t size,
+			 enum net_ip_protocol proto,
+			 s32_t timeout)
+#endif
+{
+	u32_t alloc_start = k_uptime_get_32();
+	size_t alloc_len = 0;
+	size_t hdr_len = 0;
+	struct net_buf *buf;
+
+	if (!size) {
+		return 0;
+	}
+
+	if (k_is_in_isr()) {
+		timeout = K_NO_WAIT;
```

Comment:
Should be assert instead I guess?

Reply:
no it's fine, it follows the same behavior as in net_buf API here.

Reply:
>>	if (k_is_in_isr()) {
		timeout = K_NO_WAIT;

> no it's fine, it follows the same behavior as in net_buf API here.

Yeah, but that doesn't mean that net_buf API is right doing that. Introducing new API gives a chance to tighten up API contract for the sake of more robust software. Because again, on one hand we put `(void)` in front of every memset() (MISRA), on the other, covering up such goofs a programmer may make.